### PR TITLE
feat(trace viewer): show Screenshot instead of Snapshot

### LINKED ISSUE #23964: [BUG] Canvas is not visible in the trace
### System info
- Playwright Version: [v1.35.1]
- Operating System: [macOS 13.4]
- Browser: [Chromium]
- Other info:

The content of HTML5 canvas cannot be seen in the trace, which makes debugging the tests more difficult. I am unsure if this is expected behavior or not. I would appreciate any information that can be shared on the topic.

Screenshot 2023-06-29 at 20 46 50

[trace.zip](https://github.com/microsoft/playwright/files/11908319/trace.zip)

### PR DESCRIPTION
**Pull request by Skn0tt**
Closes https://github.com/microsoft/playwright/issues/23964.

Trace snapshots are a best-effort snapshots of the browser DOM, but we can't guarantee them to be exactly what the browser showed. One example of this is `canvas` elements, where you just can't see their contents. That makes snapshots useful, but not perfect.

For those cases where the snapshot doesn't show everything, this PR introduces a new setting to show a screenshot instead. You won't be able to scroll or inspect the DOM or select a locator anymore. But if the snapshot was missing something, or displaying something wrong, you can now check the screenshot instead.

### CODE REVIEW

**Review by dgozman — COMMENTED** [review]
Let's follow up with:
- showing the click point;
- precise screencast timestamp from the browser;
- checking how this works with `snapshots: false` and `screenshots: false` traces.

File: packages/trace-viewer/src/ui/snapshotTab.tsx [hunk 1]
```diff
@@ -209,7 +229,8 @@ export const SnapshotTab: React.FunctionComponent<{
         transform: `translate(${translate.x}px, ${translate.y}px) scale(${scale})`,
       }}>
         <BrowserFrame url={snapshotInfo.url} />
-        <div className='snapshot-switcher'>
+        {(showScreenshotInsteadOfSnapshot && screencastFrame) && <img src={`sha1/${screencastFrame.sha1}`} width={screencastFrame.width} height={screencastFrame.height} />}
+        <div className='snapshot-switcher' style={showScreenshotInsteadOfSnapshot ? { display: 'none' } : undefined}>
```
[the remark this answers is not in the document]
Comment (Skn0tt) [reply packages/trace-viewer/src/ui/snapshotTab.tsx:234]: Yes that's the idiomatic way, but it has the downside of mounting & unmounting the iframes every time the setting is toggled. By using `display: none`, we keep them in the DOM. This makes all of the above logic continue to work without changes. It should also make toggling between screenshots and snapshots snappier.

File: packages/trace-viewer/src/ui/snapshotTab.tsx [hunk 2]
```diff
@@ -181,7 +201,7 @@ export const SnapshotTab: React.FunctionComponent<{
       iframe={iframeRef1.current}
       iteration={loadingRef.current.iteration} />
     <Toolbar>
-      <ToolbarButton className='pick-locator' title='Pick locator' icon='target' toggled={isInspecting} onClick={() => setIsInspecting(!isInspecting)} />
+      <ToolbarButton className='pick-locator' title={showScreenshotInsteadOfSnapshot ? 'Locators not available when showing screenshot' : 'Pick locator'} icon='target' toggled={isInspecting} onClick={() => setIsInspecting(!isInspecting)} disabled={showScreenshotInsteadOfSnapshot} />
```
[the remark this answers is not in the document]
Comment (Skn0tt) [reply packages/trace-viewer/src/ui/snapshotTab.tsx]: Agree, this should be more call-to-actiony! How do you like the new wording?

File: packages/trace-viewer/src/ui/snapshotTab.tsx [hunk 3]
```diff
@@ -154,6 +166,14 @@ export const SnapshotTab: React.FunctionComponent<{
     y: (measure.height - snapshotContainerSize.height) / 2,
   };
 
+  const page = model?.pages[0]; // TODO: figure out what to do about multiple pages.
+  const screencastFrame = React.useMemo(
+      () => snapshotInfo.timestamp && page?.screencastFrames
+        ? findClosest(page.screencastFrames, snapshotInfo.timestamp)
+        : undefined
+      , [page?.screencastFrames, snapshotInfo.timestamp]
```
[the remark this answers is not in the document]
Comment (Skn0tt) [reply packages/trace-viewer/src/ui/snapshotTab.tsx]: updated it, how do you like it now?

File: packages/trace-viewer/src/ui/snapshotTab.tsx [hunk 4]
```diff
@@ -154,6 +166,14 @@ export const SnapshotTab: React.FunctionComponent<{
     y: (measure.height - snapshotContainerSize.height) / 2,
   };
 
+  const page = model?.pages[0]; // TODO: figure out what to do about multiple pages.
```
[the remark this answers is not in the document]
Comment (Skn0tt) [reply packages/trace-viewer/src/ui/snapshotTab.tsx]: i've extracted this into a helper, but i'm still at a loss about what we should do when there's multiple pages.

**Review by Skn0tt — COMMENTED** [review]
addressed the feedback. gonna look at testing this next.

File: packages/trace-viewer/src/ui/modelUtil.ts [hunk 5]
```diff
@@ -356,6 +356,10 @@ export function prevInList(action: ActionTraceEvent): ActionTraceEvent {
   return (action as any)[prevInListSymbol];
 }
 
+export function pageForAction(action: ActionTraceEvent): PageEntry {
+  return context(action).pages[0]; // TODO: figure out what to do about multiple pages.
```
Comment (dgozman) [inline packages/trace-viewer/src/ui/modelUtil.ts:360]: We should mark each action with a page, similarly how we mark them with `contextSymbol`.

**Review by dgozman — APPROVED** [review]
Looks good to me, given that we mark each action with a page symbol.

File: packages/trace-viewer/src/ui/modelUtil.ts (hunk 5 above)
Comment (Skn0tt) [reply packages/trace-viewer/src/ui/modelUtil.ts]: I'm a little unsure if I found the right solution for this. Do you mean like in https://github.com/microsoft/playwright/pull/32248/commits/9fbf0ded351aea84ffcc1e44978673b36d5e0d2c?

Comment (Skn0tt) [reply packages/trace-viewer/src/ui/modelUtil.ts]: implemented what we discussed over video in https://github.com/microsoft/playwright/pull/32248/commits/03723b3e99bcbf61008e635a59918d737d39325f

File: packages/trace-viewer/src/entries.ts [hunk 6]
```diff
@@ -41,8 +41,10 @@ export type ContextEntry = {
 };
 
 export type PageEntry = {
+  pageId: string,
   screencastFrames: {
     sha1: string,
+    pageId: string,
```
Comment (dgozman) [inline packages/trace-viewer/src/entries.ts:47]: We don't need this one.

**Review by dgozman — APPROVED** [review]

**Comment by craigglennie** [comment]
Sorry for what's probably a dumb question: how do I use this? I've looked for an option in the UI (the one you get by passing the `--ui` flag) but failed to find anything

**Comment by Skn0tt** [comment]
Hi Craig! There's no such thing as a dumb question :D This feature got reverted, we implemented a different fix for showing canvases that'll go out with 1.49.

**Comment by craigglennie** [comment]
Thanks! I took a quick look and yay, I can see the canvas now. 

There are some visual differences from reality in the trace now, is this expected? The image we're rendering on the canvas appears to be slightly vertically stretched, and the toolbars (which are React elements, they're not in the canvas) are duplicated. That's with `1.49.0-alpha-2024-11-07`. On `1.48.2` everything looks correct (except for the original problem of not being able to see what's on the canvas).

### 1.48.2
playwright-v1-48

### 1.49.0-alpha-2024-11-07
playwright-v1-49

**Comment by Skn0tt** [comment]
This is a bug, but it's not surprising to me. We fill the canvas based on a screenshot we take during rendering, looks like we're stretching something incorrectly. Could you file a bug about this and include a repro? I'm happy to take a look.

**Pull request merged**